p2pfl/learning/callbacks/pytorch/scaffold_callback.py

```diff
@@ -0,0 +1,151 @@
+#
+# This file is part of the federated_learning_p2p (p2pfl) distribution
+# (see https://github.com/pguijas/p2pfl).
+# Copyright (c) 2022 Pedro Guijas Bravo.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, version 3.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""Callback for SCAFFOLD operations (PyTorch Lighting)."""
+
+import copy
+from typing import Any, Dict, List, Optional
+
+import lightning as pl
+import torch
+from lightning.pytorch.callbacks import Callback
+
+from p2pfl.learning.callbacks.decorators import register
+
+
+@register(callback_key='scaffold', framework='pytorch')
+class SCAFFOLDCallback(Callback):
```

Comment:
I like the approach with @register but it seems that there is a lot of duplicate code in the two SCAFFOLD callbacks. Perhaps the application of a Bridge pattern would make it easier to incorporate new frameworks with SCAFFOLD.

Remember it is also necessary to add the callback for JAX framework and others that may come later.

Reply:
I think it can introduce unnecessary complexity since tensorflow and pytorch have different hooks. For example `on_train_batch_end` receiving `tf.Variable` vs `on_before_zero_grad` receiving `torch.Tensor.`

Reply:
I leave it to your criteria @kirle, in the future you can try to reduce code, I wouldn't complicate it for now.